tiny_dnn/activations/softmax_layer.h

```diff
@@ -0,0 +1,58 @@
+/*
+    Copyright (c) 2017, Taiga Nomi
+    All rights reserved.
+
+    Use of this source code is governed by a BSD-style license that can be found
+    in the LICENSE file.
+*/
+#pragma once
+#include "tiny_dnn/activations/activation_layer.h"
+#include "tiny_dnn/layers/layer.h"
+
+namespace tiny_dnn {
+
+class softmax_layer : public activation_layer {
+ public:
+  using activation_layer::activation_layer;
+
+  std::string layer_type() const override { return "softmax-activation"; }
+
+  void forward_activation(const vec_t &x, vec_t &y) override {
+    float_t alpha = *std::max_element(x.begin(), x.end());
+    float_t denominator(0);
+    for (serial_size_t j = 0; j < x.size(); j++) {
+      y[j] = std::exp(x[j] - alpha);
+      denominator += y[j];
+    }
+
+    float_t recip_denominator = float_t(1) / denominator;
+    for (serial_size_t j = 0; j < x.size(); j++) {
+      y[j] *= recip_denominator;
```

Comment:
@Randl @karandesai-96 Doing reciprocal multiplication slightly decrease floating point precision so compilers may not do it automatically. (It depends on cases though)
http://stackoverflow.com/questions/35506226/will-the-compiler-optimize-division-into-multiplication/35506382#35506382